### PR TITLE
fix(pnp): throw `ERR_REQUIRE_ESM` when requiring an ES Module

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

Calling `require` on an ES Module throws `SyntaxError: Cannot use import statement outside a module` instead of [`ERR_REQUIRE_ESM`](https://nodejs.org/dist/latest-v16.x/docs/api/errors.html#err_require_esm)

Fixes https://github.com/yarnpkg/berry/issues/4016

**How did you fix it?**

If the `.js` file is in a module context (`package.json` contains `type: module`) throw `ERR_REQUIRE_ESM`.

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: .github/workflows/e2e-cra-workflow.yml [hunk 1]
```diff
@@ -30,6 +30,7 @@ jobs:
         yarn add -D eslint-config-react-app eslint
 
         yarn build
+        yarn test
```
Comment (merceyz) [inline .github/workflows/e2e-cra-workflow.yml:33]: This failed before this PR

File: packages/yarnpkg-pnp/sources/loader/nodeUtils.ts [hunk 2]
```diff
@@ -36,3 +37,20 @@ export function readPackage(requestPath: NativePath) {
...
+// Our error isn't as detailed since we don't have access to acorn to check
+// if the file contains ESM syntax
+export function ERR_REQUIRE_ESM(filename: string, parentPath: string | null = null) {
+  const basename = parentPath && path.basename(filename) ===
+    path.basename(parentPath) ? filename : path.basename(filename);
+
+  let msg = `require() of ES Module ${filename}${parentPath ? ` from ${
+    parentPath}` : ``} not supported.`;
+  msg += `\nInstead change the require of ${basename} in ${parentPath} to` +
+    ` a dynamic import() which is available in all CommonJS modules.`;
```
Comment (paul-soporan) [inline packages/yarnpkg-pnp/sources/loader/nodeUtils.ts:51]: These 4 lines could be a single template literal.

File: packages/yarnpkg-pnp/sources/loader/nodeUtils.ts [hunk 3]
```diff
@@ -36,3 +37,20 @@ export function readPackage(requestPath: NativePath) {
 
   return JSON.parse(fs.readFileSync(jsonPath, `utf8`));
 }
+
+// https://github.com/nodejs/node/blob/972d9218559877f7fff4bb6086afacac8933f8d1/lib/internal/errors.js#L1450-L1478
+// Our error isn't as detailed since we don't have access to acorn to check
+// if the file contains ESM syntax
+export function ERR_REQUIRE_ESM(filename: string, parentPath: string | null = null) {
+  const basename = parentPath && path.basename(filename) ===
+    path.basename(parentPath) ? filename : path.basename(filename);
```
Comment (paul-soporan) [inline packages/yarnpkg-pnp/sources/loader/nodeUtils.ts:46]: I know it's taken from the Node codebase but ending the line with `===` is quite an unusual pattern for ours.

Comment (merceyz) [reply packages/yarnpkg-pnp/sources/loader/nodeUtils.ts:46]: Indeed, our linting/formatting setup should be updated to handle this.

**Comment by swandir** [comment]
I'm curious is this problem unique to cjs loader specifically?

Does this happen because Node.js can't locate corresponding package.json for an import resolved through PnP?

**Comment by merceyz** [comment]
Yes, the issue is that it uses `internalModuleReadJSON` in a way that we can't patch so it is unable to read the zip archives. Ref https://github.com/nodejs/node/issues/33423

https://github.com/nodejs/node/blob/3743406b0a44e13de491c8590386a964dbe327bb/lib/internal/modules/cjs/loader.js#L1121
https://github.com/nodejs/node/blob/3743406b0a44e13de491c8590386a964dbe327bb/lib/internal/modules/cjs/loader.js#L329
https://github.com/nodejs/node/blob/3743406b0a44e13de491c8590386a964dbe327bb/lib/internal/modules/cjs/loader.js#L296
https://github.com/nodejs/node/blob/5f2bb885f8027e2789a1fa8679d242162b4a8633/lib/internal/modules/package_json_reader.js#L21

**Comment by swandir** [comment]
Thanks!

A somewhat similar issue affect bundlers, which too want to locate package.json (and maybe other configs) to handle advanced resolution.

In theory this kind of things should be handled via plugins providing custom resolution, but it can be tricky (https://github.com/vitejs/vite/pull/6493#discussion_r786515167) in practice.

I guess per-module resolution hooks are not completely sufficient abstraction for implementing PnP support?

I wonder whether the best approach is to try handling such edge cases in pnp bundler plugins, or rather to add an ability to read zip archives into a bundler 🤔

**Comment by merceyz** [comment]
As long as projects use `require.resolve` and `fs` to interact with stuff it should just work out of the box

**Comment by swandir** [comment]
Yes, the problem is non-js tooling like esbuild that is more common recently.

**Comment by arcanis** [comment]
Next.js fallbacks to the Node API when it detects the path [matches a pattern](https://github.com/vercel/next.js/blob/613e4c91e3fc2e2081b33c5e2da4fad4c216cbbe/packages/next/build/webpack/loaders/next-swc-loader.js#L97).

**Pull request merged**